modules/core/src/main/java/org/apache/ignite/internal/util/subscription/IterableToPublisherAdapter.java

```diff
@@ -0,0 +1,193 @@
+ * until demand is fulfilled or iterator has no more items.
+ *
+ * @param <T> The type of the entry this publisher will emit.
+ */
+public class IterableToPublisherAdapter<T> implements Publisher<T> {
+    private final Iterable<T> iterable;
+    private final Executor executor;
+    private final int batchSize;
+
+    /**
+     * Constructor.
+     *
+     * @param iterable An iterable to issue iterator for every incoming subscription.
+     * @param executor This executor will be used to drain iterator and supply entries to the subscription.
+     * @param batchSize An amount of entries to supply during a single iteration. It's always good idea
+     *      to provide some reasonable value here in order to give am ability to other publishers which share the same
+     *      executor to make progress.
+     */
+    public IterableToPublisherAdapter(Iterable<T> iterable, Executor executor, int batchSize) {
+        this.iterable = iterable;
+        this.executor = executor;
+        this.batchSize = batchSize;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber) {
+        Iterator<T> it = iterable.iterator();
+
+        Subscription subscription = new SubscriptionImpl<>(it, subscriber, executor, batchSize);
+
+        subscriber.onSubscribe(subscription);
+    }
+
+    @SuppressWarnings("FieldMayBeFinal")
+    private static class SubscriptionImpl<T> implements Subscription {
+        private static final VarHandle CANCELLED_HANDLE;
+        private static final VarHandle REQUESTED_HANDLE;
+        private static final VarHandle WIP_HANDLE;
+
+        static {
+            try {
+                Lookup lookup = MethodHandles.lookup();
+
+                CANCELLED_HANDLE = lookup.findVarHandle(SubscriptionImpl.class, "cancelled", boolean.class);
+                REQUESTED_HANDLE = lookup.findVarHandle(SubscriptionImpl.class, "requested", long.class);
+                WIP_HANDLE = lookup.findVarHandle(SubscriptionImpl.class, "wip", boolean.class);
+            } catch (NoSuchFieldException | IllegalAccessException e) {
+                throw new ExceptionInInitializerError(e);
+            }
+        }
+
+        private final Iterator<T> it;
+        private final Subscriber<? super T> subscriber;
+        private final Executor executor;
+        private final int batchSize;
+
+        private boolean cancelled = false;
+        private long requested = 0;
+        private boolean wip = false;
+
+        SubscriptionImpl(Iterator<T> it, Subscriber<? super T> subscriber, Executor executor, int batchSize) {
+            this.it = it;
+            this.subscriber = subscriber;
+            this.executor = executor;
+            this.batchSize = batchSize;
+        }
+
+        @Override
+        public void request(long n) {
+            if (n <= 0) {
+                notifyError(new IllegalArgumentException("N should be positive: n"));
```

Comment:
This message looks strange, perhaps it meant something like:

```suggestion
                notifyError(new IllegalArgumentException("N should be positive: " + n));
```

Reply:
good catch! Fixed